### ISSUE TITLE
[no-use-before-define] special case for lambdas

### ISSUE THREAD

**Issue by dpinol** [typescript-eslint/typescript-eslint#3370]
Some frameworks (such as React) favor using the syntax "const functionName = () =>" (lambdas) rather than "function functionName()".
I would like to have a way in no-use-before-define to configure functions written with lambda syntax independently from "normal variables. Maybe, apart from "variables", "functions" and "classes", there could be a new "lambdas" flag.

- [x] I have tried restarting my IDE and the issue persists.
- [x] I have updated to the latest version of the packages.
- [x] I have [read the FAQ](https://github.com/typescript-eslint/typescript-eslint/blob/master/docs/getting-started/linting/FAQ.md) and my problem is not listed.

**Repro**

```JSON
{
  "rules": {
    '@typescript-eslint/no-use-before-define': [
      'error',
      { variables: true, lambdas: false, functions: false, classes: false },
    ],  
  }
}
```

```TS
const b = () => {
  a()
}

const a = () => {
  console.log('a')
}

a()
```

**Expected Result**

I would like to be able to get the no-use-before-define error for "normal" variables, but not for those which are lambdas, since they're conceptually functions.

**Actual Result**
ESLint: 'a' was used before it was defined.(@typescript-eslint/no-use-before-define)

**Additional Info**

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `4.22.0` |
| `@typescript-eslint/parser`        | `4.19.0` |
| `TypeScript`                       | `4.2.3` |
| `ESLint`                           | `v7.25.0` |
| `node`                             | `v14.5.0` |

**Comment by dpinol** [comment]
thanks for the explanation about hoisting!
